Implement `TaskSignal` class

piotr-oles opened the pull request:
## Motivation
<img width="744" alt="image" src="https://github.com/user-attachments/assets/a0305f33-6f71-447c-8b3a-bc4a5eda36ae" />

I was surprised that this shim doesn't implement TaskSignal. It took me a while to debug why it was the case :)

## Changes
Implement `TaskSignal` class that extends `AbortSignal` class.
I use `Object.setPrototypeOf(abortSignal, TaskSignal.prototype)` to morph AbortSignal into TaskSignal.

## For follow-up
This will make possible to implement `TaskSignal.any()` method in the future.

piotr-oles committed:
Implement TaskSignal class

piotr-oles committed:
Simplify the implementation by replacing prototype

piotr-oles commented on src/task-controller.js:
```diff
@@ -17,30 +17,61 @@
-      signal.onprioritychange(e);
-    }
+    enumerable: true,
+    configurable: true,
```
This is how it's defined in native TaskSignal:
<img width="493" alt="image" src="https://github.com/user-attachments/assets/7165dc11-d8d0-4f53-ba6d-cfe907d1b4f8" />

piotr-oles commented:
Hi @shaseley and @brendankenny! 👋🏻 Could someone take a look at this PR? 🙏🏻

shaseley commented on src/task-controller.js:
```diff
@@ -17,30 +17,61 @@
-      value: null,
-      writable: true,
-      enumerable: true,
+  // Change the prototype of the signal to TaskSignal.prototype.
```
nit: You can delete this comment and the one below since it's obvious from the code itself.

piotr-oles replied:
[Remove redundant comment](https://github.com/GoogleChromeLabs/scheduler-polyfill/pull/19/commits/08a33c0313aef19f489742c3252c3118dd94390f)

shaseley reviewed (changes requested):
I think this looks pretty good, thanks for the contribution!

@brendankenny could you PTAL as well to double-check.

piotr-oles committed:
Remove redundant comment

piotr-oles commented:
Thanks for the review! Do you want me to also bump `package.version` to `1.4.0` or do you have different release process?

shaseley commented:
Hey -- really sorry for the lag. Trying to sort a few things out internally with release process/2nd reviewer.

> Thanks for the review! Do you want me to also bump `package.version` to `1.4.0` or do you have different release process?

You can leave that.

brendankenny reviewed (commented):
a few things:
- I assume any potential performance issues with `Object.setPrototypeOf` won't matter in the face of bindings, microtasks, etc overhead?
- since exposing as part of the polyfill, can you add a test (or tests) for passing it to `postTask()` (probably in `test.scheduler.js` alongside the existing `signal` tests that pulls them off a TaskController?). Also for the concern in the old comment, is it worth testing passing a polyfilled TaskSignal to another API and assert that it works correctly as an AbortSignal?
- It looks like the [existing types](https://github.com/GoogleChromeLabs/scheduler-polyfill/blob/main/types/scheduler.d.ts#L110-L123) will work for this, but have you checked? (my fault for exposing it in the types if that's what tripped you up, sorry!)

brendankenny commented:
> since exposing as part of the polyfill, can you add a test (or tests) for passing it to postTask() (probably in test.scheduler.js alongside the existing signal tests that pulls them off a TaskController?)

Oh, just kidding, brain tired. `TaskSignal` isn't constructible either, so this isn't a thing

>  Also for the concern in the old comment, is it worth testing passing a polyfilled TaskSignal to another API and assert that it works correctly as an AbortSignal?

This might still be a good idea, but the need exists for the current implementation as well, so feel free to treat it as an optional suggestion :)

piotr-oles commented:
> It looks like the [existing types](https://github.com/GoogleChromeLabs/scheduler-polyfill/blob/main/types/scheduler.d.ts#L110-L123) will work for this, but have you checked? (my fault for exposing it in the types if that's what tripped you up, sorry!)

I used these types as a reference, but I think there could be one improvement made regarding `onprioritychange` types.

This is how `onabort` is defined in `AbortSignal` type definition:
```typescript
    /** A signal object that allows you to communicate with a DOM request (such as a Fetch) and abort it if required via an AbortController object. */
    interface AbortSignal extends EventTarget {
        /**
         * Returns true if this AbortSignal's AbortController has signaled to abort, and false otherwise.
         */
        readonly aborted: boolean;
        readonly reason: any;
        onabort: null | ((this: AbortSignal, event: Event) => any);
        throwIfAborted(): void;
    }
```
Notice that it can be `null`, and it is null until you assign an event handler. Another differences are that `this` is specified as `AbortSignal` and return type is `any` instead of `void`. I would suggest changing types for `TaskSignal` to:
```diff
  class TaskSignal extends AbortSignal {
    /**
     * The priority of the signal.
     *
     * [MDN Reference](https://developer.mozilla.org/docs/Web/API/TaskSignal/priority)
     */
    readonly priority: TaskPriority;
    /**
     * Fired when the priority is changed.
     *
     * [MDN Reference](https://developer.mozilla.org/docs/Web/API/TaskSignal/prioritychange_event)
     */
-   onprioritychange: (event: TaskPriorityChangeEvent) => void;
+   onprioritychange: null | ((this: TaskSignal, event: TaskPriorityChangeEvent) => any);
  }
```
I also wonder if we should re-declare `onabort` with updated `this`:
```typescript
  class TaskSignal extends AbortSignal {
    // ...
    onabort: null | ((this: TaskSignal, event: Event) => any);
  }
```

piotr-oles commented:
> Also for the concern in the old comment, is it worth testing passing a polyfilled TaskSignal to another API and assert that it works correctly as an AbortSignal?

I will give it a try :)

piotr-oles committed:
Add test for TaskSignal and fetch compatibility

piotr-oles committed:
Update types for TaskSignal to be consistent with AbortSignal

piotr-oles commented:
I added test for `fetch()` integration and updated `TaskSignal` types. Let me know what are the  next steps :)

piotr-oles committed:
Change from `assert.okay` to `assert.ok`

There is no `assert.okay` assertion. This was working because it was called in unreachable places. It could give false negatives in try {} catch {} tests.

piotr-oles commented on test/test.taskcontroller.js:
```diff
@@ -23,7 +23,7 @@ describe('TaskController', function() {
       try {
         new TaskController('unknown');
-        assert.okay(false);
+        assert.ok(false);
```
It turns out `assert.okay` is `undefined`:
<img width="449" alt="Screenshot 2025-04-09 at 12 50 29" src="https://github.com/user-attachments/assets/e29e1356-da0f-4941-988d-b6cdb6824c65" />

It could give us false negatives in tests. If you comment one line above, the test will still pass because this will throw `Uncaught TypeError: assert.okay is not a function` and we check if `e.name === 'TypeError'`

shaseley replied:
Good catch, thanks.

piotr-oles commented:
Regarding failed `TaskSignal#constructor` test, I think I will need your help if it doesn't pass again, because it works on my machine, and the error message is very confusing. Maybe it's just because of a slower machine on CI that takes it longer to run 🤷🏻 

<img width="1041" alt="image" src="https://github.com/user-attachments/assets/b672fcb0-4e7e-4e39-9bc2-35a236607bd8" />


<img width="964" alt="image" src="https://github.com/user-attachments/assets/410aad9a-7e90-4543-8fa5-f58c8ee348fa" />

piotr-oles commented:
Can I get another look at this PR? 🙏🏻

shaseley commented on src/task-controller.js:
```diff
@@ -17,30 +17,60 @@
+
+  /**
+   * The callback to be called when the priority of the task changes.
+   * @type {Function}
```
nit: "...priority of the **signal** changes..." (the signal could be associated with multiple tasks).

piotr-oles replied:
[Make onprioritychange jsdoc more accurate](https://github.com/GoogleChromeLabs/scheduler-polyfill/pull/19/commits/19afe34295f3af9fddbc672f1f11de0cb2e4e281)

shaseley commented on src/task-controller.js:
```diff
@@ -17,30 +17,60 @@
+   * @type {string}
+   */
+  get priority() {
+    return 'user-visible';
```
It's a bit unfortunate from a readability standpoint that this is defined below as well. Maybe it's worth not returning anything here? I'd also add a comment that it's defined below and this function doesn't actually run.

We could maybe go one step further and make `makeTaskSignal` a private static method of this class to better encapsulate this.

piotr-oles replied:
Or maybe instead of storing priority in `controller.priority_`, we could use signal as a single source of truth?
We could have:
```typescript
class TaskSignal {
  /**
   * @private
   * @param {TaskController} controller
   * @param {string} priority
   */
  static make_(controller, priority) {
    Object.setPrototypeOf(controller.signal, TaskSignal.prototype);
    controller.signal.priority_ = priority;
  }

  /**
   * The priority of the task, user-visible by default.
   * @readonly
   * @type {string}
   */
  get priority() {
    /**
     * @private
     * @type {string}
     */
    return this.priority_;
  }
}

```

And then we could change `setPriority` to:
```diff
  /**
   * Change the priority of all tasks associated with this controller's signal.
   * @param {string} priority
   */
  setPriority(priority) {
    if (!SCHEDULER_PRIORITIES.includes(priority)) {
      throw new TypeError('Invalid task priority: ' + priority);
    }
    if (this.isPriorityChanging_) throw new DOMException('', 'NotAllowedError');
    if (this.signal.priority === priority) return;

    this.isPriorityChanging_ = true;

-   const previousPriority = this.priority_;
-   this.priority_ = priority;
+   const previousPriority = this.signal.priority;
+   this.signal.priority_ = priority;

    const e = new TaskPriorityChangeEvent('prioritychange', {previousPriority});
    this.signal.dispatchEvent(e);

    this.isPriorityChanging_ = false;
  }
}
```
What do you think?

shaseley replied:
Yeah that SGTM!

piotr-oles replied:
Updated: [Move priority from controller to signal](https://github.com/GoogleChromeLabs/scheduler-polyfill/pull/19/commits/4d451d62bb8dadd86e9475aed54e265b71241395)

piotr-oles replied:
I also decided to inline the TaskSignal.make_() to colocate signal-controller code:
[Inline signal morphing](https://github.com/GoogleChromeLabs/scheduler-polyfill/pull/19/commits/c976bec8b320c3f78ddd179358e1e8642542404a)

shaseley commented:
Sorry again -- I've been at conferences and summits all week, and heading OOO next week. I think I'm mostly good with this % some potential readability improvements.

piotr-oles committed:
Make onprioritychange jsdoc more accurate

piotr-oles committed:
Move priority from controller to signal

piotr-oles committed:
Inline signal morphing

shaseley reviewed (approved):
LGTM.

shaseley commented:
Hey @tunetheweb, would you be able to do the second review on this?

tunetheweb commented on src/task-controller.js:
```diff
@@ -94,19 +104,15 @@ class TaskController extends AbortController {
     this.isPriorityChanging_ = false;
 
-    makeTaskSignal(this);
+    // Morph the AbortSignal instance into a TaskSignal instance.
+    Object.setPrototypeOf(this.signal, TaskSignal.prototype);
```
My only concern is if this causes slowness as warned here:

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/setPrototypeOf

But can't say I know enough about `setPrototypeOf`, nor if this example is small enough, that it's OK here.

piotr-oles replied:
I made an arbitrary test on jsbenchmark and it seems fine. I know it's not very accurate, but it should catch any big problems: https://jsbm.dev/uq67sfUiVx8bX

tunetheweb replied:
Nice. Thanks for that.

@shaseley think we're good to merge then?

shaseley replied:
Yep!

tunetheweb reviewed (approved):
This is a very nice solution and very well implemented.

Only one small comment as to whether this will cause any slowness (I honestly don't know the answer). But other than that it LGTM.

shaseley commented:
I'll follow up about bumping the version and getting out an updated release.
